### PR TITLE
ADRs

### PR DESCRIPTION
**Pull request by saket**
Pivotal stories:
- [ADR on why we use weak references b/w SQL tables](https://www.pivotaltracker.com/story/show/160194935)
- [ADR on why the app is treated as the source of truth](https://www.pivotaltracker.com/story/show/160194943)

### CODE REVIEW

File: doc/arch/003-weak-references-between-sql-tables.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+Unfortunately, strong references work only when the data storage is centralized. Because Simple’s data is distributed across many devices, it’s impossible to guarantee the existence of all data records at the same time due to its offline by default nature. For instance, it’s easy to imagine a scenario where the app receives some blood pressures from the server, but fails to receive their associated patients because of bad network connectivity. This is expected behavior and we do not want the app to fail because of a foreign key integrity failure.
+
+## Decision
+
+SQL tables in Simple will not keep strong references using foreign keys unless they can be synced **together** in the same network call with the server. 
+
+## Consequences
+
+We are intentionally making a trade-off by letting go of data integrity on the database level. We’ll have to rely on the user interface to ensure safety against the creation of orphaned records. For instance, the UI should make it impossible to record a blood pressure if the patient does not exist.
```
Comment (pratul) [inline doc/arch/003-weak-references-between-sql-tables.md]: s/UI/user interface/

File: doc/arch/004-app-as-the-single-source-of-truth.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+## Context
+
+Unlike an online-only app that can depend on the server to create unique record IDs, Simple is designed to work completely offline. It needs to create data locally and perform two-way syncing of data with the server whenever possible. This brings two kinds of challenges,
+
+1. Resolving conflicts when merging data from the server with its local copy. 
+2. Generating IDs that are universally unique.
+
+## Decision
+
+Simple is primarily a app-only service, therefore the client app is considered as the source of truth for all [Protected Health Information (PHI)](https://en.wikipedia.org/wiki/Protected_health_information). If a conflict arises when merging a record received from the server with a local copy of the record with the same ID, the server copy is discarded if the local copy is dirty (i.e., pending a sync with the server). Otherwise, the server copy is saved.
```
Comment (pratul) [inline doc/arch/004-app-as-the-single-source-of-truth.md]: s/a app-only/an app-only/

File: doc/arch/004-app-as-the-single-source-of-truth.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+According to [Wikipedia](https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions), 
+
+> the number of random version-4 UUIDs which need to be generated in order to have a 50% probability of at least one collision is 2.71 quintillion.
+
+## Consequences
+
+1. When discarding a record received from the server in favor of its dirty local copy, any additional data in the server record will be lost. 
+
+2. Usage of UUIDs over primitives as the primary key for SQLite table has two marginal ****disadvantages: they require more storage space and are slower to query in comparison to primitives.
```
Comment (pratul) [inline doc/arch/004-app-as-the-single-source-of-truth.md]: s/marginal ****disadvantages/disadvantages/

s/storage space and are/storage space, and are/

File: doc/arch/004-app-as-the-single-source-of-truth.md (hunk 2 above)
Comment (saket) [reply doc/arch/004-app-as-the-single-source-of-truth.md]: whoops

File: doc/arch/003-weak-references-between-sql-tables.md (hunk 1 above)
Comment (saket) [reply doc/arch/003-weak-references-between-sql-tables.md]: fixed

File: doc/arch/004-app-as-the-single-source-of-truth.md (hunk 3 above)
Comment (saket) [reply doc/arch/004-app-as-the-single-source-of-truth.md]: > s/marginal ****disadvantages/disadvantages/

huh, how did this happen

> s/storage space and are/storage space, and are/

done

**Pull request merged**